pkg/controller.v1/tensorflow/pod.go

```diff
@@ -17,233 +17,49 @@ package tensorflow
 	tfConfig = "TF_CONFIG"
+	// gang scheduler name.
 
-	gangSchedulingPodGroupAnnotation = "scheduling.k8s.io/group-name"
+	//gangSchedulerName = "volcano"
+	//gangSchedulingPodGroupAnnotation = "scheduling.k8s.io/group-name"
 
 	// podTemplateRestartPolicyReason is the warning reason when the restart
 	// policy is set in pod template.
-	podTemplateRestartPolicyReason = "SettedPodTemplateRestartPolicy"
+	//podTemplateRestartPolicyReason = "SettedPodTemplateRestartPolicy"
 	// exitedWithCodeReason is the normal reason when the pod is exited because of the exit code.
-	exitedWithCodeReason = "ExitedWithCode"
+	//exitedWithCodeReason = "ExitedWithCode"
 	// podTemplateSchedulerNameReason is the warning reason when other scheduler name is set
 	// in pod templates with gang-scheduling enabled
-	podTemplateSchedulerNameReason = "SettedPodTemplateSchedulerName"
+	//podTemplateSchedulerNameReason = "SettedPodTemplateSchedulerName"
 	// podScaleDown is the normal reason when scaling down number of pods
-	podScaleDown = "PodScaleDown"
+	//podScaleDown = "PodScaleDown"
 )
 
-// reconcilePods checks and updates pods for each given TFReplicaSpec.
-// It will requeue the tfjob in case of an error while creating/deleting pods.
-func (tc *TFController) reconcilePods(
-	tfjob *tfv1.TFJob,
-	pods []*v1.Pod,
-	rtype tfv1.TFReplicaType,
-	spec *common.ReplicaSpec, rstatus map[string]v1.PodPhase) error {
-
-	// Convert TFReplicaType to lower string.
-	rt := strings.ToLower(string(rtype))
-	logger := tflogger.LoggerForReplica(tfjob, rt)
-	// Get all pods for the type rt.
-	pods, err := tc.FilterPodsForReplicaType(pods, rt)
-	if err != nil {
-		return err
-	}
-	replicas := int(*spec.Replicas)
-	restart := false
-	worker0Completed := false
-	masterRole := false
-
-	initializeTFReplicaStatuses(tfjob, rtype)
-
-	podSlices, podsToBeRemoved := tc.GetPodSlices(pods, replicas, logger)
-
-	// Scale down
-	if tfjob.Spec.EnableDynamicWorker && len(podsToBeRemoved) > 0 {
-		// Currently only allow to scale down workers
-		if rtype == tfv1.TFReplicaTypeWorker {
-			logger.Infof("Removing %d workers", len(podsToBeRemoved))
-			for _, pod := range podsToBeRemoved {
-				err := tc.PodControl.DeletePod(tfjob.Namespace, pod.Name, tfjob)
-				tc.Recorder.Eventf(tfjob, v1.EventTypeNormal, podScaleDown, "Pod: %v.%v is being removed", pod.Namespace, pod.Name)
-				if err != nil {
-					return err
-				}
-			}
-		} else {
-			logger.Warningf("Trying to scale down %s pods, which might be a mistake", rt)
-		}
-	}
-
-	for index, podSlice := range podSlices {
-		masterRole = false
-		if len(podSlice) > 1 {
-			logger.Warningf("We have too many pods for %s %d", rt, index)
-			// TODO(gaocegege): Kill some pods.
-		} else if len(podSlice) == 0 {
-			logger.Infof("Need to create new pod: %s-%d", rt, index)
-
-			// if master pod is present, select the master pod
-			// if master is not present, first worker pod is selected as the master.
-			if ContainChieforMasterSpec(tfjob) {
```

Comment:
This is TF specific. Did you change to use `isMasterRole` https://github.com/kubeflow/common/blob/f77924701cb5d8a2d6b7f49f7afaec3c6271782a/pkg/apis/common/v1/interface.go#L57

Reply:
Yes, I have implemented it in controller.go L409